### PR TITLE
changed ball moving vector formula

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by gbaranski — CHANGES_REQUESTED** [review]
stop using unsafe, instead of you can simply add another field in the World struct; handle_keys function defines a mutable access to self.

File: src/world.rs [hunk 1]
```diff
@@ -15,6 +14,7 @@ const STATS_FONT_SCALE: f32 = 1.0;
 pub struct World {
     font: Font,
     balls: [Ball; BALL_COUNT],
+    mouse_down: bool,
```
Comment (gbaranski) [inline src/world.rs]: why would you want to store the mouse state somewhere in our code?

File: src/world.rs [hunk 2]
```diff
@@ -130,19 +131,19 @@ impl World {
...
-            let vx = (dx / d).sin();
-            let vy = (dy / d).sin();
-            let v = Vec2::new(vx, vy);
-            dbg!(v);
-            primary_ball.push(-v * 2.0);
+            let d = primary_ball.position().distance(mouse_position).max(0.001);
+            let v = (*primary_ball.position() - mouse_position) / d;
+            let screen_size: f32 =
+                f64::from(screen_width().powi(2) + screen_height().powi(2)).sqrt() as f32;
+            let limit: f32 = 30.0;
```
Comment (gbaranski) [inline src/world.rs]: this should be in some constant

**Review by gbaranski — CHANGES_REQUESTED** [review]

File: src/world.rs (hunk 1 above)
Comment (ghost) [reply src/world.rs]: where should i store it?

File: src/world.rs (hunk 2 above)
Comment (ghost) [reply src/world.rs]: v and screen_size cannot be constant.
Only limit could be declared as a constant, but it won't be obvious what it refers to

File: src/world.rs (hunk 1 above)
Comment (gbaranski) [reply src/world.rs]: you don't need to store it at all. If you want to make the ball get pushed after releasing mouse button, you can use function provided by `macroquad` - `is_mouse_button_released`

File: src/world.rs (hunk 2 above)
Comment (gbaranski) [reply src/world.rs]: I was referring to `limit`

**Review by gbaranski — CHANGES_REQUESTED** [review]

File: src/world.rs [hunk 3]
```diff
@@ -130,19 +130,19 @@ impl World {
...
+            let d = primary_ball.position().distance(mouse_position).max(0.001);
+            let v = (*primary_ball.position() - mouse_position) / d;
+            let screen_size: f32 =
+                f64::from(screen_width().powi(2) + screen_height().powi(2)).sqrt() as f32;
+            let scalar = (SPEED_LIMIT * 2.0f32 * ((d as f32) / screen_size)).min(SPEED_LIMIT);
+            primary_ball.push(-v * scalar);
+        }
+
+        if is_mouse_button_released(macroquad::input::MouseButton::Left){
+            println!("t");
```
Comment (gbaranski) [inline src/world.rs]: emm

**Review by gbaranski — CHANGES_REQUESTED** [review]

File: src/world.rs [hunk 4]
```diff
@@ -130,19 +130,15 @@ impl World {
...
-            let d = primary_ball.position().distance(mouse_position);
-            let dx = primary_ball.position().x - mouse_position.x;
-            let dy = primary_ball.position().y - mouse_position.y;
-
-            let vx = (dx / d).sin();
-            let vy = (dy / d).sin();
-            let v = Vec2::new(vx, vy);
-            dbg!(v);
-            primary_ball.push(-v * 2.0);
+            let d = primary_ball.position().distance(mouse_position).max(0.001);
```
Comment (gbaranski) [inline src/world.rs]: is `.max(0.001)` really needed there?
```suggestion
            let d = primary_ball.position().distance(mouse_position);
```

**Review by gbaranski — APPROVED** [review]
after moving pushing ball logic to Ball itself and simplifying its algorithm I can say with a clear conscience that it looks good to me.

Thanks for your contribution.

File: src/world.rs (hunk 4 above)
Comment (ghost) [reply src/world.rs]: let d = primary_ball.position().distance(mouse_position).max(0.001);
            let v = (*primary_ball.position() - mouse_position) / d;
yes, to ensure d != 0
i could change it to .max(1)

**Pull request merged**